### PR TITLE
File caching alternative

### PR DESCRIPTION
**Pull request by harshad1**
In this PR I remove the cached version of File as it could cause problems.

In order to maintain high sort performance, I have added a key based comparison (inspired by how modern python does sorting).

### CODE REVIEW

**Comment by harshad1** [comment]
@gsantner This PR is ready for review. Everything seems to be working.

File: app/src/main/java/net/gsantner/opoc/util/GsUtils.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+import net.gsantner.opoc.wrapper.GsCallback;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+
+// Class for general utilities
+public class GsUtils {
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/opoc/util/GsUtils.java:14]: GsCollectionUtils?

File: app/src/main/java/net/gsantner/opoc/util/GsFileUtils.java [hunk 2]
```diff
@@ -687,85 +684,59 @@ public static String getFilenameWithTimestamp(final String... A0prefixA1postfixA
...
+     * Get a key which can be use to sort File objects
+     * This is highly performant as each file is processed exactly once.
+     * Inspired by python's sort
+     *
+     * @param sortBy   String key of what to sort
+     * @param file     The file object to get the
+     * @param dirFirst Whether to sort directories first
+     * @return A string key which can be used for comparisons / sorting
+     */
+    private static String getSortKey(final String sortBy, final File file, final boolean dirFirst) {
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/opoc/util/GsFileUtils.java:699]: get -> make?

File: app/src/main/java/net/gsantner/opoc/util/GsUtils.java [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        }
+        return out;
+    }
+
+
+    /**
+     * Apply the opertation op on each element of in, and return the result.
+     * This version runs on lists and the callback is provided with the index
+     */
+    public static <I, O> List<O> map(final List<? extends I> in, final GsCallback.r2<O, ? super I, Integer> op) {
```
[the remark this answers is not in the document]
Comment (harshad1) [reply app/src/main/java/net/gsantner/opoc/util/GsUtils.java:59]: The inplace version will require that type `O` is a child of `I`. It'll have to be a different function. Or we will have to enforce this dynamically and throw an exception or something.

File: app/src/main/java/net/gsantner/opoc/util/GsUtils.java [hunk 4]
```diff
@@ -0,0 +1,122 @@
+package net.gsantner.opoc.util;
```
[the remark this answers is not in the document]
Comment (harshad1) [reply app/src/main/java/net/gsantner/opoc/util/GsUtils.java:1]: I copied one and changed the name. Please modify as desired.

File: app/src/main/java/net/gsantner/opoc/util/GsUtils.java [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+        for (int i = 0; i < l.size(); i++) {
+            l.set(i, op.callback(l.get(i)));
+        }
+    }
+
+
+    /**
+     * Apply the opertation op on each element of in, and return the result
+     */
+    public static <I, O> List<O> map(final Collection<? extends I> in, final GsCallback.r1<O, ? super I> op) {
```
[the remark this answers is not in the document]
Comment (harshad1) [reply app/src/main/java/net/gsantner/opoc/util/GsUtils.java]: I just unified the functions

File: app/src/main/java/net/gsantner/opoc/util/GsUtils.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (harshad1) [reply app/src/main/java/net/gsantner/opoc/util/GsUtils.java]: added a separate mapInplace function to get around this

[the remark this answers is not in the document]
Comment (gsantner) [reply app/src/main/java/net/gsantner/opoc/util/GsUtils.java]: ah, out type different to in type. makes sense

**Comment by gsantner** [comment]
Sorry to say, but with the changes sorting often works not, or even the folder is not loaded.


What I have found:

* 100% always when I start the app fresh, sorting isn't applied at all seemingly (folders are not at top and i.e. QuickNote which should be alphabetically low, is on top).
* When switching sort order it somehow happens to properly sort
* when closing the app and coming back to the filemanager, the folder doesn't load at all. even when pulling down to refresh, or when opening folder again via top menu

[screenrecord-2023-03-17_19.33.54.webm](https://user-images.githubusercontent.com/6735650/225990051-68c0193c-dde2-4eb6-a648-dec865f8fbc0.webm)

screenshot-2023-03-17__19-31-25

**Comment by harshad1** [comment]
Hmmm, I am not able to reproduce any of this and it has been working consistently on my phone for quite a while now. 

Investigating

**Comment by gsantner** [comment]
Can I give some info thats helpful? I can also look more into it if I find some time on weekend

**Comment by harshad1** [comment]
- Please confirm that you are using the latest version.
- What device / OS version are you using? 
- Please check if you have any stack traces / exceptions in the logs

**Comment by harshad1** [comment]
Please also try the version I just pushed up.

**Comment by harshad1** [comment]
Pushing up another test where I use list keys instead of padded strings. All of these approaches work well for me. Just trying different things.

If you could debug and see what is happening to you then that would be great too.

File: app/src/main/java/net/gsantner/markor/activity/DocumentActivity.java [hunk 6]
```diff
@@ -209,7 +209,10 @@ private void handleLaunchingIntent(final Intent intent) {
             if (intent.hasExtra(Document.EXTRA_FILE_LINE_NUMBER)) {
                 startLine = intent.getIntExtra(Document.EXTRA_FILE_LINE_NUMBER, -1);
             } else if (intentData != null) {
-                startLine = TextViewUtils.tryParseInt(intentData.getQueryParameter("line"), -1);
+                final String line = intentData.getQueryParameter("line");
+                if (line != null) {
+                    startLine = TextViewUtils.tryParseInt(line, -1);
```
Comment (harshad1) [inline app/src/main/java/net/gsantner/markor/activity/DocumentActivity.java:214]: Unrelated fix for bug which prevented shortcuts from opening in view mode

**Review by gsantner — APPROVED** [review]
Worked better now after the changes!


Thank you, merging it 👍🏻 🎉 !

**Comment by gsantner** [comment]
@harshad1 
sorting by filesize seems to not work properly currently. I assume it's related to the key-string-sorting changes. Can you reproduce the that too?
